modules/common/tls/tls.go

```diff
@@ -0,0 +1,235 @@
+}
+
+// TLSCa contains CA-specific settings, which could be used both by services (to define their own CA certificates)
+// and by clients (to verify the server's certificate).
+type TLSCa struct {
+	// CA-specific settings
+	CaSecretName string `json:"caSecretName,omitempty"`
+}
+
+// TLS is ageneric type, which encapsulates both the service and CA configurations
+type TLS struct {
+	Service TLSService `json:"service"`
+	Ca      TLSCa      `json:"ca"`
+}
+
+// NewTLS - initialize and return a TLS struct.
+func NewTLS(clientset *kubernetes.Clientset, namespace string, service TLSService, ca TLSCa) (*TLS, error) {
+	if service.SecretName == "" {
+		service.SecretName = "default-tls-secret"
+	} else {
+		err := EnsureOrCreateSecret(clientset, namespace, service.SecretName, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if ca.CaSecretName == "" {
+		ca.CaSecretName = "default-ca-secret"
+	} else {
+		err := EnsureOrCreateSecret(clientset, namespace, ca.CaSecretName, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &TLS{
+		Service: service,
+		Ca:      ca,
+	}, nil
+}
+
+// EnsureOrCreateSecret - checks if a secret with the given name exists in the specified namespace
+func EnsureOrCreateSecret(clientset *kubernetes.Clientset, namespace, secretName string, data map[string][]byte) error {
+	_, err := clientset.CoreV1().Secrets(namespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Secret not found, create it
+			secret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secretName,
+					Namespace: namespace,
+				},
+				Data: data, // The data to store in the secret - the actual data provided by cert-manager formatted as "tls.crt": []byte("..."),  "tls.key": []byte("..."),
+			}
+			_, err := clientset.CoreV1().Secrets(namespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+			if err != nil {
+				return fmt.Errorf("Error creating secret %s: %w", secretName, err)
+			}
+		} else {
+			return fmt.Errorf("Error validating secret %s: %w", secretName, err)
+		}
+	}
+	return nil
+}
+
+// CreateVolumeMounts - add volume mount for TLS certificate and CA certificates
+func CreateVolumeMounts(tls *TLS) []corev1.VolumeMount {
+	var volumeMounts []corev1.VolumeMount
+
+	if tls.Service.SecretName != "" {
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      "tls-certs",
+			MountPath: "/etc/tls-certs",
+			ReadOnly:  true,
+			SubPath:   "tls.crt",
+		},
+			corev1.VolumeMount{
+				Name:      "tls-certs",
+				MountPath: "/etc/tls-certs",
+				ReadOnly:  true,
+				SubPath:   "tls.key",
+			})
+	}
+
+	if tls.Ca.CaSecretName != "" {
+		for i := 0; i < MaxNumberOfCAs; i++ {
+			volumeMounts = append(volumeMounts, corev1.VolumeMount{
+				Name:      fmt.Sprintf("ca-cert-%d", i),
+				MountPath: fmt.Sprintf("/etc/ca-certs/%d", i),
+				ReadOnly:  true,
+				SubPath:   fmt.Sprintf("ca-%d.crt", i),
```

Comment:
I'm not sure why, but on my env this volume mount would never expose the secret's certificate when SubPath was used. Maybe let's get rid of it and only  rely on volumes to expose a single kay "ca.crt"?

Reply:
Thanks for pointing this one. I didn't realize that when we use different mounthPath for each ca.crt, we don't have to use different names for individual CA cert.